shell/platform/fuchsia/flutter/software_surface.cc

```diff
@@ -0,0 +1,404 @@
+    fuchsia::sysmem::AllocatorSyncPtr& sysmem_allocator,
+    fuchsia::ui::composition::AllocatorPtr& flatland_allocator,
+    const SkISize& size,
+    uint32_t buffer_id) {
+  if (size.isEmpty()) {
+    FML_LOG(ERROR)
+        << "SoftwareSurface: Failed to allocate surface, size is empty";
+    return false;
+  }
+
+  fuchsia::sysmem::BufferCollectionTokenSyncPtr local_token;
+  zx_status_t allocate_status =
+      sysmem_allocator->AllocateSharedCollection(local_token.NewRequest());
+  if (allocate_status != ZX_OK) {
+    FML_LOG(ERROR) << "SoftwareSurface: Failed to allocate collection: "
+                   << zx_status_get_string(allocate_status);
+    return false;
+  }
+  fuchsia::sysmem::BufferCollectionTokenSyncPtr scenic_token;
+  zx_status_t duplicate_status = local_token->Duplicate(
+      std::numeric_limits<uint32_t>::max(), scenic_token.NewRequest());
+  if (duplicate_status != ZX_OK) {
+    FML_LOG(ERROR) << "SoftwareSurface: Failed to duplicate collection token: "
+                   << zx_status_get_string(duplicate_status);
+    return false;
+  }
+  zx_status_t sync_status = local_token->Sync();
+  if (sync_status != ZX_OK) {
+    FML_LOG(ERROR) << "SoftwareSurface: Failed to sync collection token: "
+                   << zx_status_get_string(sync_status);
+    return false;
+  }
+
+  if (session_) {
+    session_->RegisterBufferCollection(buffer_id, std::move(scenic_token));
+    buffer_id_ = buffer_id;
+  } else {
+    fuchsia::ui::composition::BufferCollectionExportToken export_token;
+    zx_status_t token_create_status =
+        zx::eventpair::create(0, &export_token.value, &import_token_.value);
+    if (token_create_status != ZX_OK) {
+      FML_LOG(ERROR)
+          << "SoftwareSurface: Failed to create flatland export token: "
+          << zx_status_get_string(token_create_status);
+      return false;
+    }
+
+    fuchsia::ui::composition::RegisterBufferCollectionArgs args;
+    args.set_export_token(std::move(export_token));
+    args.set_buffer_collection_token(std::move(scenic_token));
+    args.set_usage(
+        fuchsia::ui::composition::RegisterBufferCollectionUsage::DEFAULT);
+    flatland_allocator->RegisterBufferCollection(
+        std::move(args),
+        [](fuchsia::ui::composition::Allocator_RegisterBufferCollection_Result
+               result) {
+          if (result.is_err()) {
+            FML_LOG(ERROR)
+                << "RegisterBufferCollection call to Scenic Allocator failed";
+          }
+        });
+  }
+
+  fuchsia::sysmem::BufferCollectionSyncPtr buffer_collection;
+  zx_status_t bind_status = sysmem_allocator->BindSharedCollection(
+      std::move(local_token), buffer_collection.NewRequest());
+  if (bind_status != ZX_OK) {
+    FML_LOG(ERROR) << "SoftwareSurface: Failed to bind collection token: "
+                   << zx_status_get_string(bind_status);
+    return false;
+  }
+
+  fuchsia::sysmem::BufferCollectionConstraints constraints;
+  constraints.min_buffer_count = 1;
```

Comment:
min_buffer_count_for_camping might be set on the Scenic side, and actually there might be multiple buffers allocated for this BufferCollection. If you know that you are going to need multiple of these buffers, there can be an optimization by allocating more than one here and adding to the pool.  Good to leave as a TODO to avoid code complexity here.